llm_unlearn_ucl/unlearn_harm.py

```diff
@@ -268,23 +268,27 @@ def main(args) -> None:
             train_bad_loaders[i], train_normal_loaders[i]
         )
 
-    model.train()
-
-    # Reference model for computing KL.
+    # Pre-compute normal results for normal-loss component
+    pretrained_model_precomputed_normal_outputs_aggregated = []
     if accelerator.is_local_main_process:
-        print(f"Loading model {args.model_name} for reference ('fully learned')...")
-
-    # NOTE: accelerate.prepare only handles one model, it works now, but maybe cause some inefficiency.
-    pretrained_model = AutoModelForCausalLM.from_pretrained(
-        args.model_name,
-        cache_dir=args.cache_dir,
-        torch_dtype=torch.bfloat16,
-        trust_remote_code=True,
-    )
-    pretrained_model.to(device)
-
+        print("Precomputing the normal outputs using the pretrained model...")
+    for loader in train_normal_loaders:
+        pretrained_model_precomputed_normal_outputs_aggregated.append([])
+        for batch in loader:
+            with torch.no_grad():
+                pretrained_outputs = model(
+                    batch["input_ids"].to(device),
+                    attention_mask=batch["attention_mask"].to(device),
+                    labels=batch["labels"].to(device),
+                )
+                prob_p = torch.nn.functional.softmax(pretrained_outputs.logits, -1)
```

Comment:
Please consider moving prob_p to `cpu` to save GPU memory.

Reply:
good spot, see next commit